sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/SubscriptionPartitionLoader.java

```diff
@@ -47,15 +46,99 @@
   private final SubscriptionPath subscription;
   private final SerializableFunction<TopicPath, Integer> getPartitionCount;
   private final Duration pollDuration;
-  private final boolean terminate;
+  private final SerializableSupplier<Boolean> terminate;
+
+  private class GeneratorFn extends DoFn<Void, SubscriptionPartition> {
+    @ProcessElement
+    public ProcessContinuation processElement(
+        OutputReceiver<SubscriptionPartition> output,
+        RestrictionTracker<Integer, Integer> restrictionTracker) {
+      int previousCount = restrictionTracker.currentRestriction();
+      int newCount = getPartitionCount.apply(topic);
+      if (newCount <= previousCount) {
+        return ProcessContinuation.resume().withResumeDelay(pollDuration);
+      }
+      if (!restrictionTracker.tryClaim(newCount)) {
+        return ProcessContinuation.stop();
+      }
```

Comment:
This logic is meant to be within the restriction tracker but up to you whether you want the restriction tracker to semantically provide the logic it was built to perform.

Reply:
Understood, see above.